src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -86,6 +89,7 @@ public function supports(Request $request): ?bool
      */
     public function authenticate(RequestEvent $event)
     {
+        // @todo do we need conditionally register?
```

Comment:
I think so? I guess this is registered this way for performance? So that we don't bother doing our cookie-setting logic if there was no authentication during this request?

This is the only place now where we need the global event dispatcher now. This could potentially be done in a different way. The rough idea would be something like this:

A) Have a listener on `kernel.response` that is *always* registered when security is active (or, more accurately, when there is at least one stateful firewall)

B) Inject a service locator of the `ContextListener` for all firewalls... and something to be able to determine the current firewall (`FirewallContext` or whatever allows us to do that).

C) Then, it would look to see if the currently matched firewall has a `ContextListener`. If it does, it would call `onKernelResponse()` on it (which maybe has a different name). Internally, in `ContextListener`, instead of conditionally registering a global event listener, it could just set a flag - e.g .`$this->didAuthenticate = true` here and then reset back to  `$this->didAuthenticate = false` later (where the unregistering currently happens).

Of course, this doesn't take into account BC, but it seems like the cleaner way to do things.

Reply:
We already keep state whether or not this listener is registered: https://github.com/symfony/symfony/blob/f50e6afd7dd90e5a09e3e2ef81a2b10a8185e812/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L91

I am/was wondering whether we could use this in `onKernelResponse()` (we already do) and just return early instead of magically registering this on the fly.

Reply:
I've changed this to instead use request attributes. The current change is fully backwards compatible and the "easiest" imho.

For reference, this code is working this way to fix standalone usage of Security (in e.g. Silex). see https://github.com/symfony/symfony/pull/8946 and https://github.com/symfony/symfony/issues/8244 for some background.

Reply:
Doing dynamic registration of a listener in the main event dispatcher service as a performance optimization is likely to backfire: the event dispatcher service is optimize to compute all the sorting of listeners by priority at build time rather than during usage. But if you dynamically register a listener, it requires redoing all the sorting as it invalidates the current sorted state.